### PR TITLE
Reset Password 2: Reset Password & verify token APIs

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
## Description

Enable users to reset their password.

This PR completes **2.** 
--
~~0. Add `reset_token_digest` string attribute and `reset_token_expires_at` to users model with no restrictions.~~
~~1. Make an API for the forget password that receives the user email:
	- Finds the user by email.
	- generate a secure random **UNIQUE** complex token.
	- saves the token digest in the user `reset_token_digest` and set the `reset_token_expires_at`.
	>NOTE: The `reset_token_expires_at `= Time.now + PERIOD(seconds) where PERIOD > 0
	, PERIOD by default will be 15mins in seconds.
	- Sends the user the reset email with the token  _[Upcoming milestones]_.~~

~~>Note: For security reasons only the digest will be saved in the database and the token will be created and sent on the run.~~

~~2. Make an API for the reset password that receives the token , the new password and the user email:
	- Takes the token and validate its authenticity and integrity.
	- Validate the token expiration.
	- Update the user password.
	- Remove the user reset token attributes _[If the user update password succeeds]_.~~
	
3. Create the forget password UI and integrate it with the forget password API.
4. Create the reset password UI and integrate it with the reset password API.

### User story [Reset password]:
---
1. User navigates to sign-in page.
2. User clicks the forget password.
3. User enters the forget password UI to be prompted for their email:

>    Scenarios:
> 	1. if the user email is correct then a reset email will be sent to the user while
> 	indicating to check their inbox or spam folder.
> 	2. if the user email is not found then no email will be sent to the user but the user
> 	**WILL STILL** be indicated as if the email was sent _[Obscurity]_.

4. User receives the link with the **token digest**.
5. User _[if accesses the link before expiration]_ will be prompted for a new password and its 
confirmation.
  >Constraints: No reset password API request will initiate until:
	1. All fields are filled.
	2. Password confirmation matches the new password.
	_Note: If GLv3 include a password policy then the new password should conform to it._
6. User will get redirected to login page with the proper feedbacks and/or notifications.
[OR]
6. User will get redirected to login page if something went wrong and authenticate 
with the new password on success.

## Testing Steps

> 1. Pull the code.
> 2. Install the dependencies `bundle install && npm|yarn install`.
> 3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
> 4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:drop && rails db:create && rails db:migrate:with_data`
> 5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec && npx eslint app/javascript/* --ext .jsx,.js`
> 6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

### CODE REVIEW

**Comment by KH-Amir-TN** [comment]
# DEPENDS ON #3535

File: app/models/user.rb [hunk 1]
```diff
@@ -50,6 +78,13 @@ def generate_unique_token
     retry
   end
 
+  def invalidate_token!
+    digest = reset_digest
+    update! reset_sent_at: nil, reset_digest: nil # Remove expired/valid tokens.
+    logger.info "User(id):#{id} reset password token with digest:#{digest} got removed."
```
Comment (farhatahmad) [inline app/models/user.rb:84]: We definitely shouldn't be logging the digest, as it's considered senstiive data.

I would just `update! reset_sent_at: nil, reset_digest: nil` without any logging

File: config/application.rb [hunk 2]
```diff
@@ -28,5 +28,9 @@ class Application < Rails::Application
       record_not_found: 'Record Not Found',
       server_error: 'Something Went Wrong'
     }
+
+    # Reset token max validity period.
+    # It's advised to not increase this to more than 15 minutes.
+    config.reset_token_validity_period = 15.minutes
```
Comment (farhatahmad) [inline config/application.rb:34]: I would move this as a constant to the `User` model instead

File: app/controllers/api/v1/reset_password_controller.rb [hunk 3]
```diff
@@ -25,6 +26,44 @@ def create
...
+      # Returns: { data: Array[serializable objects] , errors: Array[String] }
+      # Does: Validates the token and reset the user password.
+
+      def reset
+        new_password = params[:user][:new_password]
+
+        return render_json status: :bad_request if new_password.blank?
+
+        # Invalidating token and changing the password.
+        @user.invalidate_token!.update! password: new_password
```
Comment (farhatahmad) [inline app/controllers/api/v1/reset_password_controller.rb:41]: I wouldn't chain these just to be safe .

```
@user.update! password: new_password
@user.invalidate_token!

```

File: app/models/user.rb [hunk 4]
```diff
@@ -31,11 +31,39 @@ def self.search(input)
...
   end
 
+  # Verifies the token existence, fetches its user and validates its expiration
+  # and invalidates the user token if expired.
+
+  def self.verify_token(token)
+    digest = generate_digest(token)
+
+    user = find_by reset_digest: digest
+    logger.info "Unable to find a user with digest: #{digest}." unless user
```
Comment (farhatahmad) [inline app/models/user.rb:41]: Same as above

File: config/application.rb (hunk 2 above)
Comment (farhatahmad) [reply config/application.rb:34]: Also - 15mins is quite short. I've normally seen it set to around 1 hour

File: app/models/user.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:84]: You're right, I wasn't thinking about it.
Will change it.

**Comment by KH-Amir-TN** [comment]
# DEPENDS ON #3555

File: app/models/user.rb [hunk 5]
```diff
@@ -66,6 +97,13 @@ def generate_activation_token!
     retry
   end
 
+  def invalidate_reset_token!
+    sent_at = reset_sent_at
+    update! reset_sent_at: nil, reset_digest: nil # Remove expired/valid tokens.
+    logger.info "User(id):#{id} reset their password reset token(sent_at):#{sent_at}."
+    self
```
Comment (farhatahmad) [inline app/models/user.rb:104]: Why are we returning `self` here?

File: app/models/user.rb [hunk 6]
```diff
@@ -32,11 +35,39 @@ def self.search(input)
...
+
+    expired = reset_token_expired?(user.reset_sent_at)
+
+    logger.info "Token(sent_at):#{user.reset_sent_at} expired for user(id):#{user.id}." if expired
+
+    user.invalidate_reset_token! if expired
+
+    return if expired
+
+    user
```
Comment (farhatahmad) [inline app/models/user.rb:58]: This should just return `true` or `false` - not raise any errors

File: app/models/user.rb [hunk 7]
```diff
@@ -32,11 +35,36 @@ def self.search(input)
...
+
+    return unless user.reset_sent_at
+
+    expired = reset_token_expired?(user.reset_sent_at)
+
+    user.invalidate_reset_token! if expired
+
+    return if expired
+
+    user
```
Comment (farhatahmad) [inline app/models/user.rb:55]: This returns `nil` and `user` not `true` and `false`

File: app/models/user.rb [hunk 8]
```diff
@@ -32,11 +35,36 @@ def self.search(input)
     all
   end
 
+  # Verifies the token existence, fetches its user and validates its expiration
+  # and invalidates the user token if expired.
+
+  def self.verify_reset_token!(token)
```
Comment (farhatahmad) [inline app/models/user.rb:41]: You can drop the `!`

File: app/models/user.rb [hunk 9]
```diff
@@ -66,6 +94,10 @@ def generate_activation_token!
     retry
   end
 
+  def invalidate_reset_token!
+    update! reset_sent_at: nil, reset_digest: nil # Remove expired/valid tokens.
+  end
+
```
Comment (farhatahmad) [inline app/models/user.rb:100]: Unless this needs to be reused, we can just move this logic into the method above

File: app/models/user.rb (hunk 7 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:55]: You prefer it to be an explicit `true|false`?

File: app/models/user.rb (hunk 8 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:41]: I'm using that as a convention just like with models'query methods.
It indicates that the method uses at least one query method that raises an exception.

I think since we want the validation operation to be silent using query methods that raises exceptions is prohibited.
Correct?

File: app/models/user.rb (hunk 9 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:100]: It's  reused in `reset_password_controller#reset` action.

File: app/controllers/api/v1/reset_password_controller.rb [hunk 10]
```diff
@@ -23,6 +24,46 @@ def create
...
+      def verify_reset_request
+        return render_json status: :bad_request unless params[:user]
+
+        token = params[:user][:token]
+
+        return render_json status: :bad_request if token.blank?
+
+        @user = User.verify_reset_token!(token)
+        render_json status: :forbidden unless @user
+      end
```
Comment (scouillard) [inline app/controllers/api/v1/reset_password_controller.rb:66]: I think that using unless and if's in the same block, back to back, makes it hard to follow (maybe it's just me).

For example, I would do `if params[:user].empty? `on line 58

Also, shouldn't we use validations for that kind of thing? If it's missing a param or if it's blank?
Or have another solution? It clutters the controller IMO.

Comment (KH-Amir-TN) [reply app/controllers/api/v1/reset_password_controller.rb:66]: Do you like using unless instead of if statements?

As for the missing it's intentional for security reasons to obscure the responses to avoid given information about sensitive operations such as the reset password.
For example if we return an error status code for an invalid token we're giving attackers a potential attack vector that can find a token in use in our system.
Another example would be on generating the token for a user that is found by email, returning an error like `:not_found` will put an endpoint to guess our users emails from the outside.

Comment (KH-Amir-TN) [reply app/controllers/api/v1/reset_password_controller.rb:66]: I think I misunderstood you, you meant we should be using the `ActionController::Parameters` API for params validation?

Comment (KH-Amir-TN) [reply app/controllers/api/v1/reset_password_controller.rb:66]: Do you like using `unless` instead of if statements?
As for the validations are you suggesting the use of the `ActionController::Parameters` API for params validation?

File: app/models/user.rb (hunk 7 above)
Comment (KH-Amir-TN) [reply app/models/user.rb:55]: This method will:
1. Generate a digest for the given the token.
2. Find a user with the corresponding digest.
3. Validate the expiration of the token.
For valid tokens the method should return the found user that's why we need it to return its record for valid tokens.
Making the method return an explicit `true` value will cause to do an extra step on the action:
1. Re-generate a digest for the given token.
2. Re-find the user by digest.
Which feels unnecessary since in the validation process it requires generating the digest and finding the user record which is guaranteed to be truthy.

File: app/controllers/api/v1/reset_password_controller.rb [hunk 11]
```diff
@@ -23,6 +24,47 @@ def create
...
+      # Does: Validates the token and reset the user password.
+
+      def reset
+        new_password = params[:user][:new_password]
+
+        return render_json status: :bad_request if new_password.blank?
+
+        # Invalidating token and changing the password.
+        # TODO: optimise this.
+        return render_json status: :internal_server_error unless @user.invalidate_reset_token
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/reset_password_controller.rb:40]: Should we make the password reset happen even when unable to invalidate the reset token?
> Note: Doing that will give the user (any one with the token) under the validity period the ability to infinitely reset the password.

File: app/models/user.rb [hunk 12]
```diff
@@ -66,6 +95,10 @@ def generate_activation_token!
     retry
   end
 
+  def invalidate_reset_token
+    update reset_sent_at: nil, reset_digest: nil # Remove expired/valid tokens.
```
Comment (KH-Amir-TN) [inline app/models/user.rb:99]: We can bypass `ActiveRecord` validations for such operations.

**Review by farhatahmad — APPROVED** [review]
